qiskit/circuit/library/standard_gates/equivalence_library.py

```diff
@@ -850,6 +850,71 @@ def _cnot_rxx_decompose(plus_ry: bool = True, plus_rxx: bool = True):
     def_swap.append(inst, qargs, cargs)
 _sel.add_equivalence(SwapGate(), def_swap)
 
+# SwapGate
+#
+# q_0: ─X─
+#       │   ≡
+# q_1: ─X─
+#
+#      ┌──────────┐┌──────┐   ┌────┐   ┌──────┐┌──────────┐┌──────┐
+# q_0: ┤ Rz(-π/2) ├┤0     ├───┤ √X ├───┤1     ├┤ Rz(-π/2) ├┤0     ├
+#      └──┬────┬──┘│  Ecr │┌──┴────┴──┐│  Ecr │└──┬────┬──┘│  Ecr │
+# q_1: ───┤ √X ├───┤1     ├┤ Rz(-π/2) ├┤0     ├───┤ √X ├───┤1     ├
+#         └────┘   └──────┘└──────────┘└──────┘   └────┘   └──────┘
+#
+q = QuantumRegister(2, "q")
+def_swap_ecr = QuantumCircuit(q)
+def_swap_ecr.rz(-pi / 2, 0)
+def_swap_ecr.sx(1)
+def_swap_ecr.ecr(0, 1)
+def_swap_ecr.rz(-pi / 2, 1)
+def_swap_ecr.sx(0)
+def_swap_ecr.ecr(1, 0)
+def_swap_ecr.rz(-pi / 2, 0)
+def_swap_ecr.sx(1)
+def_swap_ecr.ecr(0, 1)
+_sel.add_equivalence(SwapGate(), def_swap_ecr)
+
+# SwapGate
+#
+# q_0: ─X─
+#       │   ≡
+# q_1: ─X─
+#
+# global phase: 7π/4
+#      ┌─────────┐  ┌────┐        ┌────┐                         ┌────────┐»
+# q_0: ┤ Rz(π/2) ├──┤ √X ├───■────┤ √X ├──────────────────────■──┤ Rz(-π) ├»
+#      └──┬────┬─┘┌─┴────┴─┐ │ ┌──┴────┴──┐┌────┐┌──────────┐ │ ┌┴────────┤»
+# q_1: ───┤ √X ├──┤ Rz(-π) ├─■─┤ Rz(-π/2) ├┤ √X ├┤ Rz(-π/2) ├─■─┤ Rz(π/2) ├»
+#         └────┘  └────────┘   └──────────┘└────┘└──────────┘   └─────────┘»
+# «     ┌────┐┌──────────┐      ┌────┐  ┌─────────┐
+# «q_0: ┤ √X ├┤ Rz(-π/2) ├─■────┤ √X ├──┤ Rz(π/2) ├
+# «     ├────┤└──────────┘ │ ┌──┴────┴─┐└─────────┘
+# «q_1: ┤ √X ├─────────────■─┤ Rz(π/2) ├───────────
+# «     └────┘               └─────────┘
+q = QuantumRegister(2, "q")
+def_swap_cz = QuantumCircuit(q, global_phase=7 * pi / 4)
+def_swap_cz.rz(pi / 2, 0)
+def_swap_cz.sx(0)
+def_swap_cz.sx(1)
+def_swap_cz.rz(-pi, 1)
+def_swap_cz.cz(0, 1)
+def_swap_cz.sx(0)
+def_swap_cz.rz(-pi / 2, 1)
+def_swap_cz.sx(1)
+def_swap_cz.rz(-pi / 2, 1)
+def_swap_cz.cz(0, 1)
+def_swap_cz.rz(-pi, 0)
+def_swap_cz.sx(0)
+def_swap_cz.rz(-pi / 2, 0)
+def_swap_cz.rz(pi / 2, 1)
+def_swap_cz.sx(1)
+def_swap_cz.cz(0, 1)
+def_swap_cz.sx(0)
+def_swap_cz.rz(pi / 2, 0)
+def_swap_cz.rz(pi / 2, 1)
+_sel.add_equivalence(SwapGate(), def_swap_cz)
```

Comment:
Note that `rz` commutes with `cz`, so one can eliminate some `rz` gates.
I'm not sure if we can eliminate more `sx` gates.

```
def_swap_cz.rz(pi / 2, 0)
def_swap_cz.sx(0)
def_swap_cz.sx(1)
def_swap_cz.rz(pi / 2, 1)
def_swap_cz.cz(0, 1)
def_swap_cz.sx(0)
def_swap_cz.sx(1)
def_swap_cz.cz(0, 1)
def_swap_cz.rz(-pi, 0)
def_swap_cz.sx(0)
def_swap_cz.rz(-pi / 2, 0)
def_swap_cz.sx(1)
def_swap_cz.cz(0, 1)
def_swap_cz.sx(0)
def_swap_cz.rz(pi / 2, 0)
def_swap_cz.rz(pi / 2, 1)

```

Reply:
btw, why this cancellation was not done automatically by the transpiler's commutative cancellation ?

Reply:
Oh, I didn't actually test with a full transpiler run, I only ran the 2q decomposer class on it. Running a full level 3 transpile to get all the optimization passes would have been a better idea.

Reply:
Actually this decomposition has 7 `sx` gates, while it's enough to have 6 `H` gates, so we can probably improve this

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/12312/commits/f261c9fd1e96bb477f211a0a53c21a9c6d5e516d (I tweaked one of the angles to account for global phase, but otherwise it's the same cancellation)